pandas/core/series.py

```diff
@@ -3264,13 +3264,23 @@ def combine_first(self, other) -> Series:
         falcon      NaN
         dtype: float64
         """
+        from pandas.core.reshape.concat import concat
+
         new_index = self.index.union(other.index)
-        this = self.reindex(new_index, copy=False)
-        other = other.reindex(new_index, copy=False)
+
+        this = self
+        null_mask = isna(this)
+        if null_mask.any():
+            drop = this.index[null_mask].intersection(other.index[notna(other)])
```

Comment:
Can you add a quick comment what this is doing? Took me a while to figure it out, don't want to do this again in 6 months time :)

Otherwise lgtm

Reply:
i've simplified this logic a bit. should be easier to follow now